Add cs_buffer. Breaking API change.

numas13 opened the pull request:
Remove `cs_malloc()` and `cs_free()`.

`cs_disasm_iter()` is a tiny wrapper around `cs_disasm(count=1)`.

 <!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've documented or updated the documentation of every API function and struct this PR changes.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

API change unifies disassembly process. Before, there were two separate functions to disassemble one instruction at a time in a loop and many instructions into a dynamic buffer. Commit will introduce user allocatable buffer that can be used in both situations with one function.

Updating the use of `cs_disasm_iter()` is a little more complicated than `cs_disasm()`:

```c
// old api
cs_insn *insn = cs_malloc(handle);
while (cs_disasm_iter(handle, &code, &code_size, &ip, insn)) {
    disassembled_instructions += 1;
}
cs_free(insn);
```

Must be changed to:

```c
// new api
cs_buffer *buffer = cs_buffer_new(1); // create buffer with 1 element
while (cs_disasm_iter(handle, &code, &code_size, &ip, buffer)) {
    cs_insn *insn = &buffer->insn[0]; // get first insn in a buffer
    disassembled_instructions += 1;
}
cs_buffer_free(buffer); // free buffer
```

Updating the use of `cs_disasm()` is straightforward, just use `cs_buffer_new(0)` to create a buffer and pass it to `cs_disasm()`.

**Benchmarks**

```
❯ hyperfine -N ./orig/test_iter_benchmark ./test_iter_benchmark
Benchmark 1: ./orig/test_iter_benchmark
  Time (mean ± σ):     10.632 s ±  0.227 s    [User: 10.622 s, System: 0.004 s]
  Range (min … max):   10.327 s … 11.007 s    10 runs
 
Benchmark 2: ./test_iter_benchmark
  Time (mean ± σ):     10.815 s ±  0.082 s    [User: 10.812 s, System: 0.001 s]
  Range (min … max):   10.695 s … 10.986 s    10 runs
 
Summary
  './orig/test_iter_benchmark' ran
    1.02 ± 0.02 times faster than './test_iter_benchmark'
```

```
❯ ARGS="10 0x028800 0x1517dd libc.so"; hyperfine -N "./orig/test_file_benchmark $ARGS" "./test_file_benchmark $ARGS"
Benchmark 1: ./orig/test_file_benchmark 10 0x028800 0x1517dd libc.so
  Time (mean ± σ):      2.352 s ±  0.035 s    [User: 2.349 s, System: 0.002 s]
  Range (min … max):    2.299 s …  2.388 s    10 runs
 
Benchmark 2: ./test_file_benchmark 10 0x028800 0x1517dd libc.so
  Time (mean ± σ):      2.392 s ±  0.053 s    [User: 2.387 s, System: 0.003 s]
  Range (min … max):    2.329 s …  2.497 s    10 runs
 
Summary
  './orig/test_file_benchmark 10 0x028800 0x1517dd libc.so' ran
    1.02 ± 0.03 times faster than './test_file_benchmark 10 0x028800 0x1517dd libc.so'
```
 
**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Or what test cases you added. Disassembly tests should be added in suite/cs_test/issues.cs -->

Update bindings and user software.

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

Rot127 committed:
Add test with ASAN enabled.

Rot127 committed:
Fix leaks in cstool and cs.c

Rot127 committed:
Add work around so ASAN binaries don't DEADSIGNAL due to too many randomized address bits.

Rot127 committed:
Add ASAN build arguments to cstest

Rot127 committed:
Fix leaks in cstest

Rot127 committed:
Use cstest binary build by the main build.

Rot127 committed:
Add clonging step for cmocka when cstest is build

Rot127 committed:
Skip Python tests for ASAN

Rot127 committed:
Remove make build from CI

Rot127 committed:
Fix leaks in cstest.

- Rewrite split to remove leaks and improve runtime by 6%
- Add free()

Rot127 committed:
Fix cmocka external project to stable branch.

Rot127 committed:
Revert "Fix leaks in cstest."

This reverts commit bf8ee125b0c58f9c794eb081a69c80f8a71825cd.

Rot127 committed:
Fix memleaks in cstest

Rot127 committed:
Document adding of ASAN job to release guide

Rot127 committed:
Add CAPSTONE_BUILD_CSTEST to build docs

Rot127 committed:
Fix double free

Rot127 committed:
Add more detail tests to CI and fix them

Rot127 committed:
Initialize variables

Rot127 committed:
Fix typo

Rot127 committed:
Update cstest build docs

Rot127 commented on cs.c:
```diff
@@ -1093,6 +1093,83 @@ cs_err CAPSTONE_API cs_option(csh ud, cs_opt_type type, size_t value)
+void CAPSTONE_API cs_buffer_free(cs_buffer *buffer) {
+	if (buffer->insn) {
+		for (size_t i = 0; i < buffer->capacity; ++i) {
+			cs_mem_free(buffer->insn[i].detail);
```
Haven't check the changes below yet, but this probably needs a NULL check for `buffer->insn[i] != NULL`

numas13 replied:
I'm not sure. `buffer->insn` is not an array of pointers to `cs_insn`.

Rot127 replied:
You are right. Was stuck in the old implementation.

Rot127 commented on cs.c:
```diff
@@ -1093,6 +1093,83 @@ cs_err CAPSTONE_API cs_option(csh ud, cs_opt_type type, size_t value)
+		return true;
+	}
+
+	if (required < buffer->capacity) {
```
The `if (...)` can be removed, because the for loop tests it.

numas13 replied:
`buffer->insn[i].detail` can be allocated after previous run of `cs_disas()` and needs to be freed if capacity reduced.

numas13 replied:
I've refactored `cs_buffer_reserve_exact()` and fixed memory leak if `cs_mem_realloc()` fails.

Rot127 replied:
Ok, let's wait for @aquynh and @kabeor commenting.

Rot127 commented on cs.c:
```diff
@@ -1093,6 +1093,83 @@ cs_err CAPSTONE_API cs_option(csh ud, cs_opt_type type, size_t value)
+		return false;
+	}
+	if (required > buffer->capacity) {
+		ssize_t diff = required - buffer->capacity;
```
```suggestion
		size_t diff = required - buffer->capacity;
```

Rot127 commented on cs.c:
```diff
@@ -1093,6 +1093,83 @@ cs_err CAPSTONE_API cs_option(csh ud, cs_opt_type type, size_t value)
+}
+
+CAPSTONE_EXPORT
+bool CAPSTONE_API cs_buffer_reserve(cs_buffer *buffer, size_t additional) {
```
~~All these API functions still documentation please. But better wait with it until the other maintainers took a look at it.~~
Haven't seen the header files. All good.

Rot127 commented on include/capstone/capstone.h:
```diff
@@ -654,96 +663,103 @@ CAPSTONE_EXPORT
+ @return: returns false if failed to allocate memory.
+ */
+CAPSTONE_EXPORT
+bool CAPSTONE_API cs_buffer_reserve(cs_buffer *buffer, size_t additional);
```
Please document the detail that it at least reserves 16 elements.

Rot127 reviewed (changes requested):
Looks awesome! It really cleans up the API a lot. Especially internally.
Please rebase this on onto the ASAN PR from me or wait until it is merged: https://github.com/capstone-engine/capstone/pull/2313
Just so we get extra test coverage.

@aquynh @kabeor Please take a look. I am  very much in favor of this.

Rot127 commented on cs.c:
```diff
@@ -1156,85 +1224,80 @@ size_t CAPSTONE_API cs_disasm(csh ud, const uint8_t *buffer, size_t size, uint64
 
-	insn_cache = total;
+	for (; code_size && (count == 0 || buffer->count < count); ++buffer->count) {
+		if (buffer->capacity <= buffer->count && !cs_buffer_reserve(buffer, 1)) {
```
Couldn't you also reserve the buffer once, on based on some good guess how many instructions will be disassembled? So most often `code_size / 4`, sometimes `code_size / 2`.

numas13 replied:
Good idea!

Rot127 committed:
Revert "Remove make build from CI"

This reverts commit 84f7360c6da6183cd41bec0fef3e1d0a2ee49ddf.

Rot127 committed:
Make cstest only run for cmake builds.

Rot127 committed:
Add cstest job for make build.

numas13 commented:
> Please rebase this on onto the ASAN PR from me or wait until it is merged: #2313 Just so we get extra test coverage.

There is one problem in `test_arm_regression`. Is it expected?

numas13 commented:
I've updated python bindings and review is appreciated.

numas13 committed:
Add cs_buffer. Breaking API change.

Remove cs_malloc() and cs_free().

API change unifies disassembly process. Before, there were two separate
functions to disassemble one instruction at a time in a loop and many
instructions into a dynamic buffer. Commit will introduce user allocatable
buffer that can be used in both situations with one function.

cs_disasm_iter() is a tiny wrapper around cs_disasm().

Updating the use of cs_disasm_iter():

    // old api
    cs_insn *insn = cs_malloc(handle);
    while (cs_disasm_iter(handle, &code, &code_size, &ip, insn)) {
        disassembled_instructions += 1;
    }
    cs_free(insn);

Must be changed to:

    // new api
    cs_buffer *buffer = cs_buffer_new(1); // create buffer with 1 element
    while (cs_disasm_iter(handle, &code, &code_size, &ip, buffer)) {
        cs_insn *insn = &buffer->insn[0]; // get first insn in a buffer
        disassembled_instructions += 1;
    }
    cs_buffer_free(buffer); // free buffer

Updating the use of cs_disasm() is straightforward, just use
cs_buffer_new(0) to create a buffer and pass it to cs_disasm().

aquynh commented:
please avoid breaking API

numas13 commented:
I'm closing PRs. We'll think about what can be done.
